### PR TITLE
feat[bc]: move prompt cache reconciliation into llm-llamacpp

### PR DESCRIPTION
**Pull request by gianni-cor**
## 🎯 What problem does this PR solve?

- Cached turns currently depend on SDK-side delta-message and tool bookkeeping, so a stale caller view can silently append the wrong history.
- Warm tool turns can restore KV without receiving tools, leaving the chat-template grammar unarmed after the first turn.

## 📝 How does it solve it?

- Treats messages and tools on every cached addon request as the complete authoritative history, renders once, and reconciles the resulting token/media ledger by longest common prefix.
- Persists the versioned ledger inside the existing sequence-state file, uses exact trimming for attention models, and uses bounded process-local checkpoints for recurrent and hybrid divergence.
- Keeps generated reasoning resident until the next full render naturally excludes it, while preserving transactional request rollback for cancellation and failures.
- Removes the obsolete eager reasoning compactor, reasoning-specific recovery paths, `remove_thinking_from_context`, and `RuntimeStats.thinkingBlockDiscards`.
- Updates addon documentation, examples, lifecycle tests, and bumps `@qvac/llm-llamacpp` to 0.54.0.

## 🧪 How was it tested?

- `npm run build:ts`
- `npm run typecheck`
- `npm run lint:ts`
- `npm run test:types:consumer`
- Native C++ build on `qvac-dev-linux-x64` with physical GPU1.
- 122/122 cache-ledger, runtime-stat, snapshot, reasoning, and template tests passed.
- 6/6 model-backed tool-grammar, rollback, recurrent, continuous-batching, and multimodal reconciliation tests passed.
- 3/3 restored-cache corruption and rollback tests passed.
- 33/33 cache-management lifecycle tests passed after migration to committed prefill and authoritative full-history inputs.

## 💥 Breaking Changes

**BEFORE:**

```typescript
const prompt = firstTurn ? [...tools, ...newMessages] : newMessages

await model.run(prompt, {
  cacheKey,
  generationParams: { remove_thinking_from_context: true }
})
```

**AFTER:**

```typescript
const prompt = [systemMessage, ...tools, ...fullConversation]

await model.run(prompt, { cacheKey })
```

Cached callers must resend the full conversation and complete tool list on every turn. Cache files without the embedded ledger are treated as cold misses. Current SDK releases remain incompatible until the SDK follow-up adopts this contract.

## 🔌 API Changes

```typescript
// Removed from generation parameters:
remove_thinking_from_context

// Removed from runtime statistics:
thinkingBlockDiscards
```

### CODE REVIEW

File: packages/llm-llamacpp/addon/src/model-interface/TextLlmContext.cpp [hunk 1]
```diff
@@ -1166,6 +1071,13 @@ SequenceStepResult TextLlmContext::onLogitsReady(
     flushPendingUtf8ToCallback(outputCallback);
   }
 
+  // The scheduler decodes this non-terminal token on its next step. Record it
+  // provisionally now; any decode/cancel failure restores the pre-request
+  // ledger and state transactionally.
+  if (!finished && inlineDecodeBatch == nullptr) {
```
Comment (yingying0906) [inline packages/llm-llamacpp/addon/src/model-interface/TextLlmContext.cpp]: The sampled token goes into the ledger before the batcher decodes it. If a cancel lands in that gap, cancelSlotLocked syncs to currentPos without it, commits and saves. The file then has ledger positions nPast+1 under a header of nPast, and the next load fails the totals check, so that cacheKey is broken until the file is deleted. Maybe truncate the ledger to nPast on commit, and assert positions() == nPast in serialize?

File: packages/llm-llamacpp/addon/src/model-interface/TextLlmContext.cpp [hunk 2]
```diff
@@ -1114,11 +1024,6 @@ SequenceStepResult TextLlmContext::onLogitsReady(
...
       }
       reasoningState_.inside_reasoning = false;
-      // EOS substitution: the original EOS reached the capture site with
-      // capture still off and the substituted close never does, so seed it
-      // here or the restored state opens a block nothing closes.
-      compactor_.recordCloseMarkerForReplay(tokenId);
-      compactor_.requestCloseCapture();
       if (reasoningState_.cached_newline_token != LLAMA_TOKEN_NULL) {
         forcedTokens_.push_back(reasoningState_.cached_newline_token);
         forcedTokens_.push_back(reasoningState_.cached_newline_token);
```
Comment (yingying0906) [inline packages/llm-llamacpp/addon/src/model-interface/TextLlmContext.cpp]: The early return below at line 1036 skips appendResidentToken, but the scheduler still decodes the close tag. Mtmd appends it at MtmdLlmContext.cpp:1880. The ledger ends up one behind KV, and a saved file will fail to load.

**Review by yingying0906 — COMMENTED** [review]
Two things outside the diff: docs/architecture/ADDON-FUZZING.md:161 still lists ReasoningBlockCompactor, which this PR deletes. And packages/inference/src/schemas/completion-stream.ts:73 plus packages/cli/src/serve/extensions/openai/schemas/common.ts:263 still forward remove_thinking_from_context, so those requests will throw once they pin 0.54+.

File: packages/llm-llamacpp/addon/src/model-interface/MtmdLlmContext.cpp [hunk 3]
```diff
@@ -571,47 +593,105 @@ LlmContext::EvalMessageResult MtmdLlmContext::evalMessageWithTools(
...
+                    textTokenCount,
+                    pendingReuseEntries_ > ledgerEntryIndex
+                        ? pendingReuseEntries_ - ledgerEntryIndex
+                        : size_t{0})
+              : 0;
+      ledgerEntryIndex += textTokenCount;
+      if (skip == textTokenCount) {
+        continue;
+      }
+      if (skip > 0) {
```
Comment (yingying0906) [inline packages/llm-llamacpp/addon/src/model-interface/MtmdLlmContext.cpp]: This suffix loop never writes current_.pos. If a later chunk throws, e.g. an image encode OOM, restorePreRequestCacheState sees pos unchanged and trims nothing, so KV keeps cells the ledger doesn't have and the next request fails to decode. It also skips the stop check for this chunk.

File: packages/llm-llamacpp/addon/src/model-interface/TextLlmContext.cpp [hunk 4]
```diff
@@ -1281,270 +1214,256 @@ void TextLlmContext::configureReasoningTags(
...
+        break;
+      }
+    }
+    if (reuse == 0) {
+      clearSequenceMemory(modelCtx_.lctx);
+      residentLedger_.entries.clear();
+      nPast_ = 0;
+      checkpoint = "cold";
+    }
+  } else if (!needsFullStateSnapshot_ && reuseTarget < cachedLength) {
```
Comment (yingying0906) [inline packages/llm-llamacpp/addon/src/model-interface/TextLlmContext.cpp]: For SWA models without swa_full, like Gemma 3/4 and gpt-oss, the cells before reusePos may already be reused, so decoding only the suffix can run with part of the window missing and no error. llama-server checks seq_pos_min > n_past - n_swa and falls back to a full reprocess. Can we do the same here?

File: packages/llm-llamacpp/addon/src/model-interface/TextLlmContext.cpp [hunk 5]
```diff
@@ -1281,270 +1214,256 @@ void TextLlmContext::configureReasoningTags(
...
+    const std::vector<llama_token>& fullPrompt, bool isPrefillOnlyRequest) {
+  pendingPromptLedger_ = cache::fromTokens(fullPrompt);
+  const size_t prefix =
+      cache::commonPrefix(residentLedger_, pendingPromptLedger_);
+  const size_t cachedLength = residentLedger_.entries.size();
+  // A fully reused prompt has no decode step and therefore produces no fresh
+  // logits for generation. Match llama-server's cache-prompt behavior by
+  // backing up one token so the final prompt token is decoded again. A
+  // prefill-only request needs no logits and can reuse the complete prompt.
+  size_t reuseTarget = prefix;
```
Comment (yingying0906) [inline packages/llm-llamacpp/addon/src/model-interface/TextLlmContext.cpp]: For a prefill-only request there's no one-token back-off, so an exact repeat gives an empty plan. On the batch path addRequestAt returns ErrEmptyTokens and submit throws "failed to add to batch". The single path handles this fine.

File: packages/llm-llamacpp/addon/src/model-interface/TextLlmContext.cpp [hunk 6]
```diff
@@ -1281,270 +1214,256 @@ void TextLlmContext::configureReasoningTags(
...
+          checkpoint.c_str(),
+          reuse,
+          nPast_));
+
+  return std::vector<llama_token>(fullPrompt.begin() + reuse, fullPrompt.end());
 }
 
-void TextLlmContext::resetToolDefinitionsDropped() {
-  toolDefinitionsDropped_ = 0;
+void TextLlmContext::commitCacheRequest() {
```
Comment (yingying0906) [inline packages/llm-llamacpp/addon/src/model-interface/TextLlmContext.cpp]: The only checkpoint is the pre-request state, which always has the last turn's raw output. If the template re-renders that answer differently, for example by stripping thinking, no checkpoint is a prefix and hybrid models go cold every turn. Maybe also save one at the end of prefill, like llama-server?

File: packages/llm-llamacpp/addon/src/utils/SequenceStateSnapshot.cpp [hunk 7]
```diff
@@ -214,12 +175,28 @@ bool snapshotRecurrentState(
...
+  }
+
   // Write the full state (KV + recurrent) to a temp file via
   // `llama_state_seq_save_file`. Internally this calls
   // `state_seq_write_data(io, seq_id, /*flags=*/0)`, llama.cpp's
   // full-state sequence path. We do not save any prompt tokens
-  // alongside the state; those are recovered from `nPast` /
-  // `postReasoningTokens_` at restore time.
+  // alongside the state; the ledger lives in the cache transaction.
   std::string path = makeUniqueSnapshotPath(seqId);
```
Comment (yingying0906) [inline packages/llm-llamacpp/addon/src/utils/SequenceStateSnapshot.cpp]: These snapshots now get written for every cached request, up to 32 per sequence, as qvac_llamacpp_seq_<pid>_<seq>_<n>.bin in the shared temp dir with default permissions. On Linux another user can read the conversation state. Could we use a private 0700 dir or O_EXCL with mode 0600, plus a random name?

File: packages/llm-llamacpp/addon/src/model-interface/CacheLedger.hpp [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+  friend bool operator==(const Entry& a, const Entry& b) {
+    return a.kind == b.kind && a.identity == b.identity &&
+           a.positions == b.positions && a.cacheTokens == b.cacheTokens;
+  }
+};
+
+struct Ledger {
+  std::vector<Entry> entries;
+
+  [[nodiscard]] llama_pos positions(size_t end) const {
```
Comment (yingying0906) [inline packages/llm-llamacpp/addon/src/model-interface/CacheLedger.hpp]: deserialize only checks that each span is > 0, so two entries near INT32_MAX overflow this sum. Summing in int64 and rejecting anything above n_ctx would close it.

File: packages/llm-llamacpp/addon/src/model-interface/TextLlmContext.cpp [hunk 9]
```diff
@@ -1281,270 +1214,256 @@ void TextLlmContext::configureReasoningTags(
...
-void TextLlmContext::setOpenThinkSpan(llama_pos start) {
-  compactor_.setOpenSpan(start);
+void TextLlmContext::restoreCacheStateTokens(
+    const std::vector<llama_token>& tokens) {
+  const cache::DecodedLedger decoded =
+      cache::deserialize(tokens.data(), tokens.size());
+  if (decoded.nPast != decoded.cacheTokens) {
+    throw std::runtime_error("text cache has divergent position/KV totals");
+  }
+  residentLedger_ = decoded.ledger;
```
Comment (yingying0906) [inline packages/llm-llamacpp/addon/src/model-interface/TextLlmContext.cpp]: Token ids from the file aren't checked against the vocab. They don't reach the sampler today, but rejecting ids outside [0, n_vocab) here would be safer. Same in Mtmd at line 1977.

File: packages/llm-llamacpp/addon/src/model-interface/ContinuousBatchScheduler.cpp [hunk 10]
```diff
@@ -414,17 +416,8 @@ uint32_t ContinuousBatchScheduler::submitLocked(QueuedRequest&& queued) {
...
-  // but `remove_thinking_from_context` is a TextLlmContext-level toggle that
-  // sits outside `common_params`. Apply it directly to the slot driver here.
-  // No restore needed: the driver is destroyed when the slot is freed.
-  if (request.overrides.remove_thinking_from_context) {
-    driver->setRemoveThinkingFromContext(
-        *request.overrides.remove_thinking_from_context);
-  }
-
   const bool isCacheLoaded = driver->loadCache(request.cacheKey);
+  driver->setCacheReconciliationEnabled(!request.cacheKey.empty());
```
Comment (yingying0906) [inline packages/llm-llamacpp/addon/src/model-interface/ContinuousBatchScheduler.cpp]: preparePrefill below runs with mutex_ held, and for recurrent or hybrid models it can now restore a checkpoint from disk. drainFinishedLocked unlocks for the same kind of restore, as its comment explains, so admission probably needs the same.

File: packages/llm-llamacpp/addon/src/model-interface/CacheLedger.hpp [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+inline constexpr const char* CACHE_CHECKPOINTS_MAX_BYTES_KEY =
+    "cache_checkpoints_max_bytes";
+inline constexpr const char* CACHE_CHECKPOINTS_MAX_BYTES_KEY_DASHED =
+    "cache-checkpoints-max-bytes";
+inline constexpr const char* CACHE_CHECKPOINT_STORAGE_KEY =
+    "cache_checkpoint_storage";
+inline constexpr const char* CACHE_CHECKPOINT_STORAGE_KEY_DASHED =
+    "cache-checkpoint-storage";
+
+struct CheckpointPolicy {
```
Comment (yingying0906) [inline packages/llm-llamacpp/addon/src/model-interface/CacheLedger.hpp]: maxBytes defaults to 0, so validateCheckpointBudget returns early and up to 32 full snapshots per sequence are kept. On a hybrid model with a big context this can reach GBs, which worries me on mobile. Should the default have a byte cap or a smaller count? Worth checking with the team.

File: packages/llm-llamacpp/addon/src/model-interface/MtmdLlmContext.cpp [hunk 12]
```diff
@@ -1365,6 +1240,9 @@ void MtmdLlmContext::loadMedia(const std::vector<uint8_t>& media) {
             qvac_errors::general_error::InvalidArgument),
         errorMsg);
   }
+  const std::string mediaId =
```
Comment (yingying0906) [inline packages/llm-llamacpp/addon/src/model-interface/MtmdLlmContext.cpp]: With full history resent, every past image is decoded and hashed byte by byte each turn before we know it's cached. Hashing the raw input bytes instead would avoid that.

File: packages/llm-llamacpp/package.json [hunk 13]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@qvac/llm-llamacpp",
-  "version": "0.53.1",
+  "version": "0.54.0",
```
Comment (yingying0906) [inline packages/llm-llamacpp/package.json]: main already released 0.54.0, so this needs a merge from main and a bump to 0.55.0.

File: packages/llm-llamacpp/src/index.ts [hunk 14]
```diff
@@ -1427,15 +1371,7 @@ namespace LlmLlamacpp {
...
-    /**
-     * Number of `<think>` (or model-equivalent) reasoning blocks dropped
-     * from the KV cache at end-of-generation by the
-     * `remove_thinking_from_context` feature. Per-inference for single
-     * requests; summed across completed slots for batch requests. 0 when
-     * the model has no recognised reasoning channel, when the feature
-     * was disabled per-request, or when no reasoning blocks were emitted.
-     */
-    thinkingBlockDiscards: number;
+    /** Legacy counter retained for stats-shape compatibility; always 0. */
```
Comment (yingying0906) [inline packages/llm-llamacpp/src/index.ts]: This comment is left without its field, since thinkingBlockDiscards was removed. It also shows up in index.d.ts on toolDefinitionsDropped.

File: packages/llm-llamacpp/examples/warm-turn-grammar-repro.js [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+// Every cached request sends the complete conversation and tool definitions.
+// The addon renders that authoritative input once, rebuilds the grammar from
+// the same render, reuses the longest matching KV prefix, and decodes only the
+// new suffix. Resending tools therefore arms the warm-turn grammar without
+// appending a duplicate tool block.
+//
+//   MODEL=/path/to/Qwen3-1.7B-Q4_0.gguf bare warm-turn-grammar-repro.js
+//   VERBOSITY=3 shows the addon's own "tokenizeChat ... nTools=N" lines.
+
+const LlmLlamacpp = require('@qvac/llm-llamacpp')
```
Comment (yingying0906) [inline packages/llm-llamacpp/examples/warm-turn-grammar-repro.js]: The other examples use require('../index') and bare-process for exit.
